db/compaction_picker.cc

```diff
@@ -660,6 +661,22 @@ Compaction* CompactionPicker::CompactRange(
   }
   assert(output_path_id < static_cast<uint32_t>(ioptions_.cf_paths.size()));
 
+  // for bottom level compaction, skip files that are created during the
+  // current compaction
+  if (max_sst_file_number > 0) {
+    assert(input_level == output_level);
+    std::vector<FileMetaData*> inputs_shrinked;
+    for (size_t i = 0; i < inputs.size(); ++i) {
+      if (inputs[i]->fd.GetNumber() <= max_sst_file_number) {
+        inputs_shrinked.emplace_back(inputs[i]);
+      }
+    }
+    if (inputs_shrinked.empty()) {
+      return nullptr;
+    }
+    inputs.files.swap(inputs_shrinked);
```

Comment:
Is it a problem that the files may not cover a contiguous key-range after this code filters some out?

I wonder if we should pick each compaction consisting of one contiguous range of files that all need to be compacted. We can then set `compaction_end` to the end of that range and another call will be made for the next contiguous range, etc.

Reply:
Hm, it looks like the following `ExpandInputsToCleanCut` refills any holes if we pick non-contiguous files here. It seems safe, but OTOH some cases may be suboptimal. For example, the first and last file might be the only ones that need compaction (i.e., `inputs_shrinked` contains only those two). But then `ExpandInputsToCleanCut` will repopulate the input files with everything in the range, which leaves us with no files skipped.

Reply:
Thanks @ajkr !
let me see if I understood the idea correctly:
so in `CompactionPicker::CompactRange`, after the call to `GetOverlappingInputs`, we will still filter out new files larger than `max_sst_file_number`, then maybe call a new function to break up `inputs_shrinked` into continuous ranges of files, and for each range, call `SetupOtherInputs` and then create/register one new Compaction? 
This will create a number of `Compaction` objects but the caller `DBImpl::RunManualCompaction` seems to expect only one `Compaction`?

Reply:
I guess if it's only suboptimal in some corner cases but optimizes in general cases, we can live with that. Or if really needed, we can also do some trick in `ExpandInputsToCleanCut` but that might be a bit error-prone

Reply:
Rather than leaving holes in the middle, should we just cut the compaction?
If we have files [a b] [c d] [e f], and [c d] is newly generated, should we pick [a b] to compact itself, and then [e f] in a following one?